### PR TITLE
[Wallet] Improve handling of recent recipients

### LINKED ISSUE #361: Recent recipients don't update until app restart
### Expected Behavior

Recents work and update while app is open.

### Current Behavior

Recent recipient cache should be in redux store but currently just live in Send.tsx
Because of this. Updates to recents aren't reflected until we re-import contacts (e.g. restart the app).

### LINKED ISSUE #394: App should not persist recipient cache to Redux Async Storage
### Expected Behavior

Not persisted 

### Current Behavior

Persisted but ignored on rehydrate (too much wasted storage)

### PR DESCRIPTION
**Pull request by jmrossy**
### Description

- Improve handling of recent recipients (simplify logic)
- Fix recent recipients not showing up til app restart
- Blacklist recipientCache from persist by creating a new recipients reducer
- Reorganize recipient related code and files
- Fetch balances after wallet import

### Tested

Sent to all diff recipient types
Sent to the same recip more than once to confirm recent deduping.

### Related issues

- Fixes #361 
- Fixes #394 
- Fixes #396 

### Backwards compatibility

Yes but users's existing recent recipients will be wiped clean

recents-fix

### CODE REVIEW

File: packages/mobile/src/send/Send.tsx [hunk 1]
```diff
@@ -81,24 +92,8 @@ const mapDispatchToProps = {
   importContacts,
 }
 
-// TODO(Rossy) move this into redux as I've done for the full cache (don't forget to blacklist it when you do)
-let recentCache: Recipient[] | null = null
-
-const THROTTLE_TIME = 50
-
-const defaultRecipientPhoneNumber = '+10000000000'
```
Comment (jmrossy) [inline packages/mobile/src/send/Send.tsx:89]: Just moved these consts up to top

File: packages/mobile/src/recipients/actions.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+export enum Actions {
+  SET_RECIPIENT_CACHE = 'SEND/SET_RECIPIENT_CACHE',
+}
+
+export interface SetRecipientCacheAction {
+  type: Actions.SET_RECIPIENT_CACHE
+  recipients: NumberToRecipient
+}
+
+export type ActionTypes = SetRecipientCacheAction
```
Comment (jeanregisser) [inline packages/mobile/src/recipients/actions.ts:12]: nit: I made a note we're mixing up `Actions` and `ActionTypes`.

File: packages/mobile/src/send/saga.ts [hunk 3]
```diff
@@ -52,6 +52,8 @@ export async function getSendFee(
 
 export function* watchQrCodeDetections() {
   while (true) {
+    // TODO(Rossy) this gets called taken multiple times before a user can press the send button
```
Comment (jeanregisser) [inline packages/mobile/src/send/saga.ts:55]: 👍

File: packages/mobile/src/utils/insertAtBeginning.ts [hunk 4]
```diff
@@ -1,4 +1,10 @@
 // inserts the new item at the first index and makes sure it's not present twice
 export const insertAtBeginning = <T>(item: T, items: T[]) => {
-  return [item, ...items.filter((anItem) => anItem !== item)]
+  return [
+    item,
+    ...items.filter(
+      (existingItem) =>
+        existingItem !== item && JSON.stringify(existingItem) !== JSON.stringify(item)
```
Comment (jeanregisser) [inline packages/mobile/src/utils/insertAtBeginning.ts]: This method works generically but seems a bit heavyweight. I wonder if there was a simpler `key` to use for the data we have?

**Review by jeanregisser — DISMISSED** [review]
Awesome work!

File: packages/mobile/src/utils/insertAtBeginning.ts (hunk 4 above)
Comment (jmrossy) [reply packages/mobile/src/utils/insertAtBeginning.ts]: Good point, I'll find a better way to do this

File: packages/mobile/src/recipients/recipient.ts [hunk 5]
```diff
@@ -13,10 +12,10 @@ import Logger from 'src/utils/Logger'
 const TAG = 'utils/recipient'
```
Comment (martinvol) [inline packages/mobile/src/recipients/recipient.ts]: Forgot this tag. Feels like the name and folder shouldn't be the same.

Comment (jmrossy) [reply packages/mobile/src/recipients/recipient.ts]: Good catch!

**Review by martinvol — DISMISSED** [review]

**Review by martinvol — APPROVED** [review]

**Pull request merged**